### DISCUSSION TITLE
Building a Maven p2 repo with features

### DISCUSSION THREAD

**Discussion by reckart** [eclipse-tycho/tycho#931] [Q&A] [answered]
This is a kind of a follow-up to https://github.com/eclipse/tycho/discussions/925 attempting to build the update site as a Maven update site.

I am trying to adapt the approach taken by the [Jetty team](https://github.com/eclipse/jetty.project/blob/75f5cc68e4358decc58e4ba7ba00b8f7f5f9bd32/jetty-p2/pom.xml) and have added this `tycho-p2-repository-plugin` configuration to my update site Maven module:

```
      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>tycho-p2-repository-plugin</artifactId>
        <version>${tycho-version}</version>
        <executions>
          <execution>
            <id>maven-p2-site</id>
            <phase>prepare-package</phase>
            <goals>
              <goal>assemble-maven-repository</goal>
            </goals>
            <configuration>
              <includeReactor>true</includeReactor>
              <includeDependencies>false</includeDependencies>
            </configuration>
          </execution>
        </executions>
      </plugin>
```

However, there are two problems with that:

1)  it only picks up the OSGi bundles from the reactor, not the feature modules (that are built using Tycho)
2) it only works if building the entire reactor, not when only building the update site module.

So I tried adjusting to adding the features as dependencies and including dependencies in the build:

```
  <dependencies>
    <dependency>
      <groupId>org.apache.uima</groupId>
      <artifactId>org.apache.uima.runtime.feature.group</artifactId>
      <version>3.3.0-SNAPSHOT</version>
    </dependency>
    <dependency>
      <groupId>org.apache.uima</groupId>
      <artifactId>org.apache.uima.tools.feature.group</artifactId>
      <version>3.3.0-SNAPSHOT</version>
    </dependency>
  </dependencies>

      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>tycho-p2-repository-plugin</artifactId>
        <version>${tycho-version}</version>
        <executions>
          <execution>
            <id>maven-p2-site</id>
            <phase>prepare-package</phase>
            <goals>
              <goal>assemble-maven-repository</goal>
            </goals>
            <configuration>
              <includeReactor>false</includeReactor>
              <includeDependencies>true</includeDependencies>
            </configuration>
          </execution>
        </executions>
      </plugin>
```

But that yields an empty update site again:

```
% unzip -l uimaj-eclipse-update-site/target/p2-site.zip
Archive:  uimaj-eclipse-update-site/target/p2-site.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
      403  05-04-2022 12:11   content.xml
      610  05-04-2022 12:11   artifacts.xml
---------                     -------
     1013                     2 files
```

Also, the `uimaj-eclipse-update-site` contains a `category.xml` file, but it seems to be ignored, although I can see it being added via the `-categoryDefinition` to the `org.eclipse.tycho.p2.tools.publisher.TychoFeaturesAndBundlesPublisher` when build with `-X`.

I can see an odd error message printed to stderr, but it doesn't tell me which file is the problem:

```
    Generating metadata for ..
    Generation completed with success [0 seconds].
>>  [Fatal Error] :1:1: Premature end of file.
    [INFO] Building zip: .../uimaj-eclipse-update-site/target/p2-site.zip
```

**Comment by laeubi** [comment, marked as answer]
> 1. it only picks up the OSGi bundles from the reactor, not the feature modules (that are built using Tycho)

I think features are not supported right now neverless.

> 2. it only works if building the entire reactor, not when only building the update site module.

That is intentional as the goal for jetty is to include everything currently build without any further configuration, if you only build the site, then there is nothing to include (how should the mojo know?)

> So I tried adjusting to adding the features as dependencies and including dependencies in the build

That is the correct approach, but will only work for bundles right now, so instead of a feature you need to add a dependency to all the bundles that feature includes.

By the way, this feature was specifically crafted to be used in "non tycho" cases, it does not work well if you try to mix things here, see
- #436

**Reply by reckart** [reply]
The Jetty p2 site artifacts on Maven Central do not contain bundle JARs, but when I try the same build (with `<includeReactor>true` (or alternatively with `<includeDependencies>true` and having bundle dependencies in the POM`, then I get a p2 site ZIP that includes the bundle JARs:

```
% unzip -l p2-site.zip
Archive:  p2-site.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  05-04-2022 13:33   plugins/
    69393  05-04-2022 13:33   plugins/org.apache.uima.caseditor.ide_3.3.0.SNAPSHOT.jar
   384297  05-04-2022 13:33   plugins/org.apache.uima.caseditor_3.3.0.SNAPSHOT.jar
    41398  05-04-2022 13:33   plugins/org.apache.uima.launcher_3.3.0.SNAPSHOT.jar
   477198  05-04-2022 13:33   plugins/org.apache.uima.desceditor_3.3.0.SNAPSHOT.jar
    86961  05-04-2022 13:33   plugins/org.apache.uima.pear_3.3.0.SNAPSHOT.jar
  4442963  05-04-2022 13:33   plugins/org.apache.uima.runtime_3.3.0.SNAPSHOT.jar
    10933  05-04-2022 13:33   plugins/org.apache.uima.debug_3.3.0.SNAPSHOT.jar
    15386  05-04-2022 13:33   plugins/org.apache.uima.jcas.jcasgenp_3.3.0.SNAPSHOT.jar
    63825  05-04-2022 13:33   content.xml
     7209  05-04-2022 13:33   artifacts.xml
---------                     -------
  5599563                     11 files
```

What's different in Jetty that they do not get the JARs embedded in the P2 Maven site? I thought not having the bundle JARs in there was kind of the whole point of the `assemble-maven-repository` goal?

**Reply by laeubi** [reply]
> The Jetty p2 site artifacts on Maven Central do not contain bundle JARs

That's on purpose as all references to the artifacts are deployed maven URLs you will see this when looking into the `artifacts.xml`

> but when I try the same build

The same as jetty?

> I get a p2 site ZIP that includes the bundle JARs:

Its hard to tell without seeing the full build-script in action. e.g. the `assemble-maven-repository` is meant for packaging type `pom` if you use it with other packaging types they might interfere as more mojos are executed as you might assume.

**Reply by reckart** [reply]
I have build two update site modules now to test the "Maven p2" and the "plain p2" approaches:

* https://github.com/apache/uima-uimaj/blob/dbe58d15b8bcf3ab003f820f59729765ed4312f6/uimaj-maven-p2-site/pom.xml (Maven p2)
* https://github.com/apache/uima-uimaj/blob/dbe58d15b8bcf3ab003f820f59729765ed4312f6/uimaj-eclipse-update-site/pom.xml (plain p2)

The "maven p2" uses `pom` packaging, so nothing should interfere here (I hope).  They inherit their basic tycho config from a profile further up in the reactor hierarchy:

* https://github.com/apache/uima-uimaj/blob/dbe58d15b8bcf3ab003f820f59729765ed4312f6/uimaj-parent/pom.xml#L403

I have studied the debug output of the build, but I didn't see any hint why indicating why the "maven p2" build embeds the bundle JARs in the p2 site archive.

So I'm pondering now which way to go...

* **Maven P2**
    * 👍 No duplicate artifacts on Maven and in update site since update site directly refers to Maven artifacts
    * 👍 No separate checksums and signing of the artifacts in P2 site required
    * 👎 JARs embedded when they shoudln't
    * 👎 Features apparently not supported yet
    * 👎 `category.xml` file seems to be completely ignored and does not end up in the site archive
* **Plain P2**
    * 👍 Features supported
    * 👎 JARs embedded
    * 👎 Need to unpack the generated site, sign and hash all the artifacts, then package back together again 

Also, there's a lot of new distracting messages in the build since trying Tycho:

```
% mvn clean install -DskipTests
[INFO] Scanning for projects...
[WARNING] Could not transfer metadata org.apache.uima:parent-pom:15-SNAPSHOT/maven-metadata.xml from/to ${eclipseP2RepoId} (https://download.eclipse.org/releases/2018-12/): Cannot access https://download.eclipse.org/releases/2018-12/ with type p2 using the available connector factories: BasicRepositoryConnectorFactory
[WARNING] No explicit target runtime environment configuration. Build is platform dependent.
[WARNING] No explicit target runtime environment configuration. Build is platform dependent.
[WARNING] No explicit target runtime environment configuration. Build is platform dependent.
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[INFO] Fetching p2.index from https://download.eclipse.org/releases/2018-12/ (122B)
[INFO] Adding repository https://download.eclipse.org/releases/2018-12
[INFO] Fetching p2.index from https://download.eclipse.org/technology/epp/packages/2018-12/ (142B)
[INFO] Fetching p2.index from https://download.eclipse.org/releases/2018-12/201812191000/ (142B)
[INFO] Resolving dependencies of MavenProject: org.apache.uima:org.apache.uima.runtime.feature.group:3.3.0-SNAPSHOT @ .../git/uima-uimaj/uimaj-eclipse-feature-runtime/pom.xml
[INFO] The following requirements are not satisfied yet and must be provided through pom dependencies:
[INFO]    - org.eclipse.equinox.p2.iu; org.apache.uima.runtime [3.3.0,3.3.1)
[INFO] Resolving dependencies of MavenProject: org.apache.uima:org.apache.uima.tools.feature.group:3.3.0-SNAPSHOT @ .../git/uima-uimaj/uimaj-eclipse-feature-tools/pom.xml
[INFO] The following requirements are not satisfied yet and must be provided through pom dependencies:
[INFO]    - org.eclipse.equinox.p2.iu; org.apache.uima.jcas.jcasgenp 0.0.0
[INFO]    - org.eclipse.equinox.p2.iu; org.apache.uima.debug [3.3.0,3.3.1)
[INFO]    - org.eclipse.equinox.p2.iu; org.apache.uima.desceditor [3.3.0,3.3.1)
[INFO]    - org.eclipse.equinox.p2.iu; org.apache.uima.jcas.jcasgenp [3.3.0,3.3.1)
[INFO]    - org.eclipse.equinox.p2.iu; org.apache.uima.pear [3.3.0,3.3.1)
[INFO]    - org.eclipse.equinox.p2.iu; org.apache.uima.caseditor [3.3.0,3.3.1)
[INFO]    - org.eclipse.equinox.p2.iu; org.apache.uima.caseditor.ide [3.3.0,3.3.1)
[INFO]    - org.eclipse.equinox.p2.iu; org.apache.uima.launcher [3.3.0,3.3.1)
[WARNING] The following locally built units have been used to resolve project dependencies:
[WARNING]   org.apache.uima.runtime.feature.group/3.3.0.202205021942
[WARNING]   org.apache.uima.runtime.feature.jar/3.3.0.202205021942
[INFO] Resolving dependencies of MavenProject: org.apache.uima:uimaj-eclipse-update-site:3.3.0-SNAPSHOT @ .../git/uima-uimaj/uimaj-eclipse-update-site/pom.xml
[INFO] The following requirements are not satisfied yet and must be provided through pom dependencies:
[INFO]    - org.eclipse.equinox.p2.iu; org.apache.uima.tools.feature.group [3.3.0,3.3.1)
[WARNING] The following locally built units have been used to resolve project dependencies:
[WARNING]   org.apache.uima.runtime.feature.group/3.3.0.202205021942
[WARNING]   org.apache.uima.runtime.feature.jar/3.3.0.202205021942
```

**Reply by laeubi** [reply]
> The "maven p2" uses `pom` packaging, so nothing should interfere here (I hope). They inherit their basic tycho config from a profile further up in the reactor hierarchy:

If you use "maven p2" you should not inherit anything from tycho! That's the purpose of the mojo, that it works without any Tycho configuration! So if you are having only plain maven that should be enough.

> I have studied the debug output of the build, but I didn't see any hint why indicating why the "maven p2" build embeds the bundle JARs in the p2 site archive.

This feature is quite new, and there is also a chance that there are bugs, the best would be to start with an absolute minimal example and
provide an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue with that.

For example for the reactor part we have the following integration-test: https://github.com/eclipse/tycho/tree/master/tycho-its/projects/p2mavensite.reactor

> * `category.xml` file seems to be completely ignored and does not end up in the site archive

The category file is used if places at the default location and [you can configure its location](https://www.eclipse.org/tycho/sitedocs/tycho-p2/tycho-p2-repository-plugin/assemble-maven-repository-mojo.html#categoryFile) if you want, but the file itself is *never* part of the update-site. But you can only configure the content there, it won't pull anything in automatically! Again here, a minimal example would help to find out if there is a bug.

**Reply by reckart** [reply]
I have had a look at the tycho ITs, but they all seem to be the Java APIs and I am not even remotely familiar with those. That said, here is a reasonably minimal (I hope) example that illustrates my issue.

[tycho-p2-site-test.zip](https://github.com/eclipse/tycho/files/8621795/tycho-p2-site-test.zip)

* Run it using `mvn clean install`
* Then `unzip -l test-site/target/p2-site.zip` and you should see the JAR in the ZIP:
```
Archive:  test-site/target/p2-site.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  05-04-2022 17:19   plugins/
     2645  05-04-2022 17:19   plugins/test.bundle_0.0.1.SNAPSHOT.jar
     3115  05-04-2022 17:19   content.xml
     1381  05-04-2022 17:19   artifacts.xml
---------                     -------
     7141                     4 files
```

You can probably also see the spurious `[Fatal Error] :1:1: Premature end of file.` line in the build output.

**Reply by reckart** [reply]
Wrt. the `category.xml` file never being part of an update site - it seems when I build the features and update site using Tycho (see [here](https://github.com/eclipse/tycho/discussions/925#discussioncomment-2677959) in the other thread), then the `category.xml` file is included in the site archive when I listed its contents:

```
     1698  05-03-2022 07:16   category.xml
```
Full archive file list [here](https://github.com/eclipse/tycho/discussions/925#discussioncomment-2677959).

**Reply by laeubi** [reply]
> You can probably also see the spurious `[Fatal Error] :1:1: Premature end of file.` line in the build output.

Yeah I sometimes noticed that but no idea what it is about and it never caused issues so just ignore it :-)

> here is a reasonably minimal (I hope) example that illustrates my issue

Thanks I'll try to take a look at this.

**Reply by laeubi** [reply]
> Wrt. the `category.xml` file never being part of an update site - it seems when I build the features and update site using Tycho (see [here](https://github.com/eclipse/tycho/discussions/925#discussioncomment-2677959) in the other thread), then the `category.xml` file is included in the site archive when I listed its contents:

Tycho includes some resource, but technically a `category.xml` has no meaning for an updatesite (aka: Eclipse never reads it).

**Reply by reckart** [reply]
I think I've understood now the relation between the `project` folder and the test classes in the ITs project, but unfortunately Eclipse hangs when trying to import Tycho into the workspace.... sry.

**Reply by laeubi** [reply]
This seems to be a regression from Tycho 2.7.0 > 2.7.1 (even though I don't see a reason why), with 2.7.0 no bundles are included.

**Reply by reckart** [reply]
Btw. after trying to massage Eclipse into accepting the Tycho projects, I'm giving up on trying to import the codebase into Eclipse now... m2e claims loops in the build graph and neither clean/updating nor restarting Eclipse nor first building everything on the command line seems to help.

**Build cycle message on tycho-its** (although similar messages are on most other modules as well)
> One or more cycles were detected in the build path of project 'tycho-its'. The paths towards the cycle and cycle are:
tycho-its->{org.eclipse.tycho.core.shared, sisu-equinox-launching, tycho-core, tycho-testing-harness}
tycho-its->{org.eclipse.tycho.core.shared, sisu-equinox-launching, tycho-core, org.eclipse.tycho.p2.resolver.shared}
tycho-its->{sisu-equinox-launching, tycho-core, org.eclipse.tycho.p2.tools.shared}

**Reply by laeubi** [reply]
Yeah this is a known issue with m2e that sadly sometimes happens but should hopefully be fixed with the next release.

**Reply by reckart** [reply]
I notice another downside of the Maven-P2 approach is that the URL to the site-zip that gets deployed to the Maven repository cannot be entered into Eclipse as an update site URL. Eclipse is unable to handle the fact that it is a ZIP archive despite being able to handle ZIP archives when they are available locally. I tried it with one of [the Jetty P2 sites from Maven Central](https://repo.maven.apache.org/maven2/org/eclipse/jetty/jetty-p2/10.0.9/jetty-p2-10.0.9-p2site.zip).

**Reply by reckart** [reply]
Even when using the local `repository` folder that the maven-p2 plugin produces as an update site in Eclipse, it does not work. The bundles are displayed but when I actually try to install them, there are errors that `mvn` is not a supported protocol.

<img width="680" alt="Screenshot 2022-05-04 at 19 30 22" src="https://user-images.githubusercontent.com/1410238/166745318-e006b30c-d5bc-4a2c-b6f9-1e94ed6328aa.png">

**Reply by laeubi** [reply]
You need the latest m2eclipe installed to use maven urls.

**Reply by reckart** [reply]
Later than what comes with a fresh Eclipse 2022-03? I am really **that** bleeding edge here?!

<img width="538" alt="Screenshot 2022-05-04 at 19 33 12" src="https://user-images.githubusercontent.com/1410238/166745846-3752ff76-bd44-4680-b238-3acc1aefe96c.png">

**Reply by laeubi** [reply]
1.20 should be okay so what exact URL are you using? I haven't checked this lately ... can you make sure m2e is loaded e.g. can you create a maven project from the wizzard?

**Reply by reckart** [reply]
**Note:** Little trick that works even with ZIPs that are on Maven Central... manually weak the ZIP link to look like this: `jar:https://repo.maven.apache.org/maven2/org/eclipse/jetty/jetty-p2/10.0.9/jetty-p2-10.0.9-p2site.zip!/`

... and when with m2e 2.0.0 from the **SNAPSHOT** update site, it even installs bundles using the `mvn` protocol. 🍾 🥳

**Reply by laeubi** [reply]
> **ote:** Little trick that works even with ZIPs that are on Maven Central.

The downside is that Eclipse will download the zip multiple times!

**Reply by reckart** [reply]
Well, an Update Site containing only metadata should hopefully not be so large that it matters?

**Reply by laeubi** [reply]
I checked this and mvn protocol support was added back in Nov 25, 2020 so ther must be something broken or wrong URL used somewhere...

**Reply by laeubi** [reply]
> Well, an Update Site containing only metadata should hopefully not be so large that it matters?

Sure, just wanted to note that in case one uses that for larger ones, anyways *inside* it is still using mvn urls so actually it should work at from the very beginning without any hacks.

**Reply by reckart** [reply]
I tried the `jar:https://repo.maven.apache.org/maven2/org/eclipse/jetty/jetty-p2/10.0.9/jetty-p2-10.0.9-p2site.zip!/` URL once with a fresh Eclipse 2022-03, once after updating to the latest release version from the m2e release update side and once with the snapshot update site - it only worked after I had installed m2e from the snapshot site 🤷

**Reply by laeubi** [reply]
You must use `mvn:org.eclipse.jetty:jetty-p2:10.0.9:zip:p2site` as the repo URL.

**Comment by laeubi** [comment]
Can you please try out the current [tycho snapshot](https://github.com/eclipse/tycho/wiki#getting-tycho-snapshots) build for version `2.7.3-SNAPSHOT` if that fixes the problem for you?

**Reply by reckart** [reply]
Looks good 👍
